text/0080-datetime-extension.md

```diff
@@ -0,0 +1,367 @@
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm localized to their time.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+Note that this is identical to the previous policy. The calling application must adjust `workdayStart` and `workdayEnd`.
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1, 7].context.now.dayOfWeek
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day == 29 &&
+  principal.birthDate.month == 2 &&
+  context.now.day == 29 && 
+  context.now.month == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.now.timestamp > datetime("20200131T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to millisecond accuracy, and `duration` which represents a duration of time.
+To construct and manipulate these types we will provide the functions listed below.
+All of this functionality will be hidden behind a `datetime` feature flag (analogous to the current decimal and IP extensions), allowing users to opt-out.
+
+- `datetime(string)` constructs a datetime value. Like with other extension function constructors, strict validation requires `string` to be a string literal, although evaluation/authorization support any string-typed expression. The string must be of one of the forms, and regardless of the timezone offset is always normalized to UTC:
+  - `"YYYY-MM-DDThh:mm:ssZ"` (UTC)
+  - `"YYYY-MM-DDThh:mm:ss.SSSZ"` (UTC with millisecond precision)
+  - `"YYYY-MM-DDThh:mm:ss(+/-)hhmm"` (With timezone offset in hours and minutes)
+  - `"YYYY-MM-DDThh:mm:ss.SSS(+/-)hhmm"` (With timezone offset in hours and minutes and millisecond precision)
+- `.add(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can be used with comparison operators:
```

Comment:
I'm not sure that we want to support operator overloading here.  

Note that we don't support operator overloading for other extension types (such as decimals).

We can keep all of these operators, but for consistency of the language design, I suggest that we make them into methods:

* `DT1.lessThan(DT2)` 
* ...

And same for datetime arithmetic:

* `DT1.sub(DT2)` or whatever we want to call this method.

Question:  can durations be negative?

Reply:
How hard would it be to support overloading? The ergonomic, and readability improvements are substantial... and they're just longs, really. :)

There's no reason durations can't be negative, I think?

Reply:
Note: If we do overloading here, we should really do it for `Decimal()` numbers too.

Reply:
I don't think it would be too much effort to support (at least in the Rust/Lean implementation), we just don't have a precedent for it. If we decide to use `<` et al for date/time, then we should go back and do the same for decimal.

Also: Note that the decimal extension currently uses `.lessThan()`, so we'd either need to choose a new name for date/time or support overloading for that operation.

Reply:
I think we would have negative durations if we have something like "2024.08.20" - "2024.08.30". 

Or do we intend the operator to return the absolute value here?

If so, that's another argument for not overloading `-` for this operator.

Reply:
I think we should have negative durations and not implicitly the absolute value.

Cedar already supports some kinds of overloading, e.g., `in Group::"A"` and `in [Group::"A"]` are both valid.  And `has` works with both entities and records on its LHS.

Reply:
I guess in summary, I think overloading here (and for `Decimal`, retroactively) is OK.

Reply:
It feels to me that the order of the operands in `DT1 - DT2` may hold a lot of information about the intent of the policy. 

`(startDate - now) > 7days` vs `(now - startDate) > 7days` expresses two completely separate concepts. One asks "are you more than 7 days from starting?" The other "did you start at least 7 days ago?"

If you take absolute durations:

`abs(startDate - now) > 7days` and `abs(now - startDate) > 7days` say the same exact thing. You'd need an additional `startDate > now` or `now > startDate` to get the original intent.

Reply:
Update: we've added sections on operator overloading and negative durations to the "unresolved questions" section. Please move conversation to comments on those sections!